back/app/controllers/omniauth_callback_controller.rb

```diff
@@ -27,7 +46,24 @@ def auth_callback(verify:, authver_method:)
 
     @identity = Identity.find_or_create_with_omniauth(auth, authver_method)
 
-    @user = @identity.user || User.find_by_cimail(user_attrs.fetch(:email))
+    @user = @identity.user
+
+    if @user.nil?
+      @user = User.find_by_cimail(user_attrs.fetch(:email))
+      if @user && !authver_method.can_merge?(@user, user_attrs)
```

Comment:
Good that you're delegating this to the actual method :+1: 

I was thinking that we could generalize this even further and also let the authver method decide how an existing user is matched. E.g. the current hard coded `find_by_cimail` could become something like `find_matching_user?(user_attrs)`. It would let us achieve the same thing as `can_merge?` (when nothing is returned, there is no match and we error the same way you do now) but would give more flexibility to the method to match in different ways, which would help us with methods that don't support email, for example (we've had requests for this).

Just sharing this as an aside, if this is tested and properly validated and there are other concerns I'm not thinkin about, then I'm fine with this sticking to this approach.

Reply:
Indeed, it would be simpler, but it seems we could not handle all the cases with `find_matching_user`.

E.g., if `find_matching_user` returns `nil`, what does it mean? That we cannot merge and we should show the message about it? Or that we should create a new user with this email?

Currently, if `find_by_cimail` returns `nil`, we create a new user.

```
   find_by_cimail
  /               \
nil                existing user
 |                        |
(1)
we create             can_merge?
a new user           /         \
                    (2)       (3)
             yes, we merge       no, we show error message about names


    find_matching_user
  /               \
 (1)
nil                existing user
 |                        |
                         (2)
we create             we merge
a new user
```

So, there are three cases, and it seems `find_matching_user` can handle only two. But maybe I'm missing something.

Reply:
Yes, you're right. We could still do it through return values from `find_matching_user`, letting the method decide between these 3 cases (and maybe there'll be more in the future even). But I'd consider that out of scope for now.